### ISSUE TITLE
Use of `PATH_MAX` on WASI

### ISSUE THREAD

**Issue by sunfishcode** [cptpcrd/realpath-ext#2]
857a424201a2f7add8bfe634ebd9d0f264c69ca5 introduced this code for WASI:
```rust
#[cfg(target_os = "wasi")]
const PATH_MAX: usize = 4096;
```
On WASI, paths aren't limited to 4096 bytes; the specific path length is host-dependent, and not something that can be known at compile time, with today's WASI.

Would it be possible to revise the code to use a loop that reallocates until the buffer is big enough?

**Comment by cptpcrd** [comment]
> Would it be possible to revise the code to use a loop that reallocates until the buffer is big enough?

I thought about this a bit, and came to some conclusions:

- It sounds like changing the default behavior of `realpath()` on WASI is the Right Thing To Do.
- I don't want to change the default behavior on non-WASI, because on most other OSes, failing with `ENAMETOOLONG` is generally the correct thing to do.
- However, if I'm going to be adding functionality to allow retrying with a bigger buffer, I want to somehow expose this functionality more generally.

I think the best way to solve all 3 of these concerns is to follow the Rust idiom and add a "builder" which provides the functionality of `realpath()`, but also allows setting various options. In this case, the option might be something like "can the path exceed `PATH_MAX`?", which would default to `true` on WASI, and `false` everywhere else.

`realpath()` would then be modified to use this "builder," so its default behavior would change on WASI. `realpath_raw()` would also get a "builder", though it would be slightly different to accommodate `realpath_raw()`'s different API. 

I'll start putting these changes together.

**Comment by sunfishcode** [comment]
I would like to propoose that for realpath-ext, on non-WASI, returning `ENAMETOOLONG` for strings over `PATH_MAX` is not the most desirable thing to do.

It makes realpath-ext less useful in situations where code is prepared to handle strings longer than `PATH_MAX`, such as code that parses path strings into components, code that knows how to handle long paths using sequences of `openat` calls, or debugging code that just prints the string out for users to see.

And even in the case where the string returned from realpath-ext is immediately passed to a system call which only accepts up to `PATH_MAX` bytes, there's nothing wrong with passing it longer strings, it'll fail gracefully. And when it does fail, application code is then more likely to provide a useful error message containing the path to the end user.

**Comment by cptpcrd** [comment]
> I would like to propoose that for realpath-ext, on non-WASI, returning `ENAMETOOLONG` for strings over `PATH_MAX` is not the most desirable thing to do.
> 
> It makes realpath-ext less useful in situations where code is prepared to handle strings longer than `PATH_MAX`, such as code that parses path strings into components, code that knows how to handle long paths using sequences of `openat` calls, or debugging code that just prints the string out for users to see.
> 
> And even in the case where the string returned from realpath-ext is immediately passed to a system call which only accepts up to `PATH_MAX` bytes, there's nothing wrong with passing it longer strings, it'll fail gracefully. And when it does fail, application code is then more likely to provide a useful error message containing the path to the end user.

I've thought about this, and ultimately realized that it is a moot point. Generally speaking, `realpath-ext` _cannot_ handle paths longer than `PATH_MAX`, because it needs to make syscalls (namely `readlink()` and `stat()`) which themselves are constrained by `PATH_MAX`. I verified this with some quick testing.

In fact, `realpath-ext` can only be made to validly return paths longer than `PATH_MAX` if a) it was passed a relative path and b) the combination of the current directory + the (resolved) relative path exceeds `PATH_MAX`. I think that this is a case which arguably justifies the inclusion of a option to bypass the `PATH_MAX` check, but I do not think it justifies changing the _default_ behavior.

(Side note: `realpath-ext` _could_ be made to handle paths longer than `PATH_MAX` more generally by using `openat` like you described. However, I'm not going to be changing `realpath()` to do this because it be fairly complicated and add significant extra overhead. I may add this functionality to `realpath-ext` eventually, but it would probably be in the form of a separate function within this crate.)

**Comment by cptpcrd** [comment]
I've pushed some changes (mostly in ba60ea05e993492a5fd23bb8e7e701bfd67c9dc5) that add the "builders" I described, and loosen the `PATH_MAX` limitation on WASI. Note that there still is a length limit on WASI, mainly to prevent possible "infinite loop" scenarios. I've set it to 32768 by default, which should be enough to resolve any sane path.

I'm planning on publishing a new release with these changes tomorrow unless something comes up.
